File: c++/74. Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty()) return false;
        int m = matrix.size(), n = matrix[0].size();
        if (n == 0) return false;
        
        int l = 0, r = m-1, mid;
        while (l <= r) {
            mid = (l+r)>>1;
            if (matrix[mid][0] == target)
                return true;
            else if (matrix[mid][0] > target)
                r = mid - 1;
            else
                l = mid + 1;
        }
        
        int row = max(0, r);
        l = 0; r = n-1;
        while (l <= r) {
            mid = (l+r)>>1;
            if (matrix[row][mid] == target)
                return true;
            else if (matrix[row][mid] > target)
                r = mid - 1;
            else
                l = mid + 1;
        }
        
        return false;
    }
};

File: python/74. Search a 2D Matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix: return False
        row = len(matrix)-1
        col = 0
        while row >=0 and col<len(matrix[0]):
            if matrix[row][col] > target:
                row -= 1
            elif matrix[row][col] < target:
                col += 1
            else:
                return True
        return False

# 思路: 选取左下角或者右上角作为起始点,
# 时间复杂度:O(m+n)
